pyro/infer/svgd.py

```diff
@@ -0,0 +1,198 @@
+import math
+from contextlib import ExitStack  # python 3
+
+import torch
+from torch.distributions import biject_to
+
+import pyro
+from pyro.distributions import Delta
+from pyro.infer.trace_elbo import Trace_ELBO
+from pyro.infer.autoguide.guides import AutoDelta
+from pyro.infer.autoguide.initialization import init_to_sample
+from pyro.distributions.util import sum_rightmost
+
+
+def vectorize(fn, num_particles, max_plate_nesting):
+    def _fn(*args, **kwargs):
+        with pyro.plate("num_particles_vectorized", num_particles, dim=-max_plate_nesting - 1):
+            return fn(*args, **kwargs)
+    return _fn
+
+
+class _SVGDGuide(AutoDelta):
+    """
+    This modification of :class:`AutoDelta` is used internally in the
+    :class:`SVGD` inference algorithm.
+    """
+    def __init__(self, model):
+        super(_SVGDGuide, self).__init__(model, prefix="svgd", init_loc_fn=init_to_sample)
+
+    def __call__(self, *args, **kwargs):
+        """
+        An automatic guide with the same ``*args, **kwargs`` as the base ``model``.
+
+        :return: A dict mapping sample site name to sampled value.
+        :rtype: dict
+        """
+        # if we've never run the model before, do so now so we can inspect the model structure
+        if self.prototype_trace is None:
+            self._setup_prototype(*args, **kwargs)
+
+        plates = self._create_plates()
+        result = {}
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            with ExitStack() as stack:
+                for frame in site["cond_indep_stack"]:
+                    if frame.vectorized:
+                        stack.enter_context(plates[frame.name])
+                value = pyro.param("{}_{}".format(self.prefix, name),
+                                   site["value"].detach(),
+                                   constraint=site["fn"].support)
+                # The following lines are where this method differs from the AutoDelta version.
+                # For SVGD (but not MAP) we need the log det jacobian terms for correctness.
+                transform = biject_to(site["fn"].support)
```

Comment:
@fritzo please let me know if you have any suggestions for how to clean this part up. this machinery is a bit opaque to me

Reply:
@fritzo pending any suggestions on this bit, i'd like to get this PR merged. this is because i think i'm going to have to refactor things a bit anyway to support a cleaner/more performant/more general version in #1991. that will also help me tackle any lingering shape issues (e.g. scalar random variables). lmk if you have any questions

Reply:
TODO: please try to factor this code and share some code with pyro/infer/autoguide/guides.py. Maybe add a helper in autoguides/guide.py. This is very tricky code and just had a bug fixed a few weeks ago. If we copy and paste, the two bits will diverge.